common/nexus/incoming_service_registry.go

```diff
@@ -0,0 +1,355 @@
+	persistence p.NexusIncomingServiceManager,
+	logger log.Logger,
+) *IncomingServiceRegistry {
+	return &IncomingServiceRegistry{
+		status:           common.DaemonStatusInitialized,
+		config:           config,
+		serviceDataReady: future.NewFuture[struct{}](),
+		services:         make(map[string]*nexus.IncomingService),
+		matchingClient:   matchingClient,
+		persistence:      persistence,
+		logger:           logger,
+	}
+}
+
+func (r *IncomingServiceRegistry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	refreshCtx := headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemBackgroundCallerInfo,
+	)
+	r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.refreshServicesLoop)
+}
+
+func (r *IncomingServiceRegistry) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	r.refreshPoller.Cancel()
+	<-r.refreshPoller.Done()
+}
+
+func (r *IncomingServiceRegistry) Get(ctx context.Context, id string) (*nexus.IncomingService, error) {
+	if !r.serviceDataReady.Ready() {
+		if err := r.initializeServices(ctx); err != nil {
+			r.logger.Error("error initializing Nexus incoming service cache.", tag.Error(err))
+			return nil, serviceerror.NewInternal("error loading Nexus incoming service data.")
+		}
+	}
+
+	r.dataLock.RLock()
+	defer r.dataLock.RUnlock()
+
+	service, ok := r.services[id]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("could not find Nexus incoming service with ID: %v", id))
+	}
+
+	return service, nil
+}
+
+func (r *IncomingServiceRegistry) refreshServicesLoop(ctx context.Context) error {
+	minWaitTime := r.config.refreshMinWait()
+
+	for ctx.Err() == nil {
+		start := time.Now()
+		_ = backoff.ThrottleRetryContext(ctx, r.refreshServices, r.config.refreshRetryPolicy, nil)
+		elapsed := time.Since(start)
+
+		// In general, we want to start a new call immediately on completion of the previous
+		// one. But if the remote is broken and returns success immediately, we might end up
+		// spinning. So enforce a minimum wait time that increases as long as we keep getting
+		// very fast replies.
+		if elapsed < minWaitTime {
+			common.InterruptibleSleep(ctx, minWaitTime-elapsed)
+			// Don't let this get near our call timeout, otherwise we can't tell the difference
+			// between a fast reply and a timeout.
+			minWaitTime = min(minWaitTime*2, r.config.refreshLongPollTimeout()/2)
+		} else {
+			minWaitTime = r.config.refreshMinWait()
+		}
+	}
+
+	return ctx.Err()
+}
+
+// refreshServices sends long-poll requests to matching to check for any updates to service data.
+// It waits to send any requests until service data has been initialized.
+func (r *IncomingServiceRegistry) refreshServices(ctx context.Context) error {
+	// Wait for service data to be initialized before long polling.
+	if _, err := r.serviceDataReady.Get(ctx); err != nil {
+		return err
+	}
```

Comment:
This should be checked at the top of `refreshServicesLoop`.

Reply:
My thinking was it's better to have it here so that if there is an error waiting for the future, the poll will be retried on the next loop iteration.